File: platform/polyglot/gemini_openai_sdk/py/gemini_openai_sdk/tools.py
"""
Tools Module - Tool Definitions and Executors

OpenAI function-calling compatible tool definitions and execution.
"""

import ast
import json
import operator
from typing import Any, Callable, Dict, List, Optional

from .logger import create

logger = create("gemini_openai_sdk", __file__)


# =============================================================================
# Tool Definitions (OpenAI Function Calling Format)
# =============================================================================

WEATHER_TOOL: Dict[str, Any] = {
    "type": "function",
    "function": {
        "name": "get_weather",
        "description": "Get the current weather for a location",
        "parameters": {
            "type": "object",
            "properties": {
                "location": {
                    "type": "string",
                    "description": "The city and state, e.g., San Francisco, CA",
                },
                "unit": {
                    "type": "string",
                    "enum": ["celsius", "fahrenheit"],
                    "description": "Temperature unit",
                },
            },
            "required": ["location"],
        },
    },
}

CALCULATOR_TOOL: Dict[str, Any] = {
    "type": "function",
    "function": {
        "name": "calculate",
        "description": "Perform a mathematical calculation",
        "parameters": {
            "type": "object",
            "properties": {
                "expression": {
                    "type": "string",
                    "description": "The mathematical expression to evaluate",
                },
            },
            "required": ["expression"],
        },
    },
}

# Default tools list
DEFAULT_TOOLS: List[Dict[str, Any]] = [WEATHER_TOOL, CALCULATOR_TOOL]


# =============================================================================
# Tool Executors
# =============================================================================

def execute_weather(args: Dict[str, Any]) -> Dict[str, Any]:
    """
    Simulated weather tool implementation.

    Args:
        args: Dictionary with 'location' and optional 'unit'

    Returns:
        Weather data dictionary
    """
    logger.debug("execute_weather: enter", args=args)

    location = args.get("location", "Unknown")
    unit = args.get("unit", "celsius")

    result = {
        "location": location,
        "temperature": 22 if unit == "celsius" else 72,
        "unit": unit,
        "conditions": "sunny",
        "humidity": 45,
    }

    logger.info("execute_weather: success", location=location, unit=unit)
    return result


_SAFE_OPS = {
    ast.Add: operator.add,
    ast.Sub: operator.sub,
    ast.Mult: operator.mul,
    ast.Div: operator.truediv,
    ast.USub: operator.neg,
    ast.UAdd: operator.pos,
}


def _eval_math_node(node):
    """Recursively evaluate an AST node containing only arithmetic."""
    if isinstance(node, ast.Constant) and isinstance(node.value, (int, float)):
        return node.value
    if isinstance(node, ast.BinOp):
        op = _SAFE_OPS.get(type(node.op))
        if op is None:
            raise ValueError(f"Unsupported operator: {type(node.op).__name__}")
        return op(_eval_math_node(node.left), _eval_math_node(node.right))
    if isinstance(node, ast.UnaryOp):
        op = _SAFE_OPS.get(type(node.op))
        if op is None:
            raise ValueError(f"Unsupported operator: {type(node.op).__name__}")
        return op(_eval_math_node(node.operand))
    raise ValueError(f"Unsupported expression: {type(node).__name__}")


def safe_eval_math(expression: str):
    """Safely evaluate a mathematical expression using AST parsing."""
    tree = ast.parse(expression.strip(), mode="eval")
    return _eval_math_node(tree.body)


def execute_calculate(args: Dict[str, Any]) -> Dict[str, Any]:
    """
    Calculator tool implementation.

    Uses AST-based safe math evaluation (no eval/exec).

    Args:
        args: Dictionary with 'expression' key

    Returns:
        Calculation result dictionary
    """
    logger.debug("execute_calculate: enter", args=args)

    expression = args.get("expression", "")

    # Security: Only allow safe math characters
    allowed_chars = set("0123456789+-*/.() ")
    if not all(c in allowed_chars for c in expression):
        logger.warn("execute_calculate: unsafe expression rejected", expression=expression)
        return {"expression": expression, "error": "Invalid expression - unsafe characters"}

    try:
        result = safe_eval_math(expression)
        logger.info("execute_calculate: success", expression=expression, result=result)
        return {"expression": expression, "result": result}
    except Exception as e:
        logger.error("execute_calculate: failed", expression=expression, error=str(e))
        return {"expression": expression, "error": "Invalid expression"}


# =============================================================================
# Tool Registry
# =============================================================================

_tool_registry: Dict[str, Callable[[Dict[str, Any]], Dict[str, Any]]] = {
    "get_weather": execute_weather,
    "calculate": execute_calculate,
}


def register_tool(
    name: str,
    executor: Callable[[Dict[str, Any]], Dict[str, Any]],
    definition: Optional[Dict[str, Any]] = None,
) -> None:
    """
    Register a custom tool.

    Args:
        name: Tool function name
        executor: Function that executes the tool
        definition: Optional OpenAI tool definition
    """
    logger.info("register_tool: registering", name=name)
    _tool_registry[name] = executor


def execute_tool(function_name: str, arguments: Dict[str, Any]) -> Dict[str, Any]:
    """
    Execute a registered tool by name.

    Args:
        function_name: Name of the tool function
        arguments: Arguments dictionary for the tool

    Returns:
        Tool execution result
    """
    logger.debug("execute_tool: enter", function_name=function_name)

    executor = _tool_registry.get(function_name)

    if not executor:
        logger.error("execute_tool: unknown function", function_name=function_name)
        return {"error": f"Unknown function: {function_name}"}

    try:
        result = executor(arguments)
        logger.debug("execute_tool: success", function_name=function_name)
        return result
    except Exception as e:
        logger.error("execute_tool: failed", function_name=function_name, error=str(e))
        return {"error": str(e)}


def process_tool_calls(tool_calls: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """
    Process multiple tool calls from model response.

    Args:
        tool_calls: List of tool call objects from API response

    Returns:
        List of tool results with call info and execution results
    """
    logger.debug("process_tool_calls: processing %d calls", len(tool_calls))
    results = []

    for tool_call in tool_calls:
        func_name = tool_call["function"]["name"]
        func_args_str = tool_call["function"]["arguments"]

        try:
            func_args = json.loads(func_args_str)
        except json.JSONDecodeError:
            func_args = {}
            logger.warn("process_tool_calls: failed to parse arguments", func_name=func_name)

        tool_result = execute_tool(func_name, func_args)

        results.append({
            "id": tool_call.get("id"),
            "function": func_name,
            "arguments": func_args,
            "result": tool_result,
        })

    logger.info("process_tool_calls: processed %d calls", len(results))
    return results


def get_available_tools() -> List[str]:
    """Get list of registered tool names."""
    return list(_tool_registry.keys())

File: platform/polyglot/gemini_openai_sdk/mjs/gemini-openai-sdk/tools.mjs
/**
 * Tools Module - Tool Definitions and Executors
 *
 * OpenAI function-calling compatible tool definitions and execution.
 */

import { create } from './logger.mjs';

const logger = create('gemini-openai-sdk', import.meta.url);

// =============================================================================
// Tool Definitions (OpenAI Function Calling Format)
// =============================================================================

export const WEATHER_TOOL = {
  type: 'function',
  function: {
    name: 'get_weather',
    description: 'Get the current weather for a location',
    parameters: {
      type: 'object',
      properties: {
        location: {
          type: 'string',
          description: 'The city and state, e.g., San Francisco, CA',
        },
        unit: {
          type: 'string',
          enum: ['celsius', 'fahrenheit'],
          description: 'Temperature unit',
        },
      },
      required: ['location'],
    },
  },
};

export const CALCULATOR_TOOL = {
  type: 'function',
  function: {
    name: 'calculate',
    description: 'Perform a mathematical calculation',
    parameters: {
      type: 'object',
      properties: {
        expression: {
          type: 'string',
          description: 'The mathematical expression to evaluate',
        },
      },
      required: ['expression'],
    },
  },
};

export const DEFAULT_TOOLS = [WEATHER_TOOL, CALCULATOR_TOOL];

// =============================================================================
// Tool Executors
// =============================================================================

/**
 * Simulated weather tool implementation.
 * @param {object} args - Arguments with location and optional unit
 * @returns {object} Weather data
 */
export function executeWeather(args) {
  logger.debug('executeWeather: enter', { args });

  const { location = 'Unknown', unit = 'celsius' } = args;

  const result = {
    location,
    temperature: unit === 'celsius' ? 22 : 72,
    unit,
    conditions: 'sunny',
    humidity: 45,
  };

  logger.info('executeWeather: success', { location, unit });
  return result;
}

/**
 * Tokenize a math expression into numbers and operators.
 * @param {string} expr
 * @returns {string[]}
 */
function tokenizeMath(expr) {
  const tokens = [];
  let i = 0;
  while (i < expr.length) {
    if (expr[i] === ' ') { i++; continue; }
    if ('+-*/()'.includes(expr[i])) {
      tokens.push(expr[i++]);
    } else if (/[0-9.]/.test(expr[i])) {
      let num = '';
      while (i < expr.length && /[0-9.]/.test(expr[i])) num += expr[i++];
      tokens.push(num);
    } else {
      throw new Error(`Unexpected character: ${expr[i]}`);
    }
  }
  return tokens;
}

/**
 * Safely evaluate a math expression using recursive descent parsing.
 * Supports +, -, *, /, parentheses, and unary +/-.
 * @param {string} expression
 * @returns {number}
 */
function safeEvalMath(expression) {
  const tokens = tokenizeMath(expression);
  let pos = 0;
  const peek = () => tokens[pos];
  const consume = () => tokens[pos++];

  function parseExpr() {
    let left = parseTerm();
    while (peek() === '+' || peek() === '-') {
      const op = consume();
      const right = parseTerm();
      left = op === '+' ? left + right : left - right;
    }
    return left;
  }

  function parseTerm() {
    let left = parseFactor();
    while (peek() === '*' || peek() === '/') {
      const op = consume();
      const right = parseFactor();
      left = op === '*' ? left * right : left / right;
    }
    return left;
  }

  function parseFactor() {
    if (peek() === '(') {
      consume();
      const val = parseExpr();
      if (peek() !== ')') throw new Error('Missing closing parenthesis');
      consume();
      return val;
    }
    if (peek() === '-') { consume(); return -parseFactor(); }
    if (peek() === '+') { consume(); return parseFactor(); }
    const token = consume();
    if (token === undefined) throw new Error('Unexpected end of expression');
    const num = parseFloat(token);
    if (isNaN(num)) throw new Error(`Invalid number: ${token}`);
    return num;
  }

  const result = parseExpr();
  if (pos !== tokens.length) throw new Error('Unexpected token after expression');
  return result;
}

/**
 * Calculator tool implementation using safe evaluation.
 * @param {object} args - Arguments with expression
 * @returns {object} Calculation result
 */
export function executeCalculate(args) {
  logger.debug('executeCalculate: enter', { args });

  const { expression = '' } = args;

  // Security: Only allow safe math operations
  const allowedChars = new Set('0123456789+-*/.() ');
  const isSafe = [...expression].every(c => allowedChars.has(c));

  if (!isSafe) {
    logger.warn('executeCalculate: unsafe expression rejected', { expression });
    return { expression, error: 'Invalid expression - unsafe characters' };
  }

  try {
    const result = safeEvalMath(expression);
    logger.info('executeCalculate: success', { expression, result });
    return { expression, result };
  } catch (err) {
    logger.error('executeCalculate: failed', { expression, error: err.message });
    return { expression, error: 'Invalid expression' };
  }
}

// =============================================================================
// Tool Registry
// =============================================================================

const toolRegistry = new Map([
  ['get_weather', executeWeather],
  ['calculate', executeCalculate],
]);

/**
 * Register a custom tool.
 * @param {string} name - Tool function name
 * @param {Function} executor - Function that executes the tool
 */
export function registerTool(name, executor) {
  logger.info('registerTool: registering', { name });
  toolRegistry.set(name, executor);
}

/**
 * Execute a registered tool by name.
 * @param {string} functionName - Name of the tool function
 * @param {object} args - Arguments for the tool
 * @returns {object} Tool execution result
 */
export function executeTool(functionName, args) {
  logger.debug('executeTool: enter', { functionName });

  const executor = toolRegistry.get(functionName);

  if (!executor) {
    logger.error('executeTool: unknown function', { functionName });
    return { error: `Unknown function: ${functionName}` };
  }

  try {
    const result = executor(args);
    logger.debug('executeTool: success', { functionName });
    return result;
  } catch (err) {
    logger.error('executeTool: failed', { functionName, error: err.message });
    return { error: err.message };
  }
}

/**
 * Process multiple tool calls from model response.
 * @param {Array} toolCalls - List of tool call objects from API response
 * @returns {Array} List of tool results
 */
export function processToolCalls(toolCalls) {
  logger.debug('processToolCalls: processing', { count: toolCalls.length });
  const results = [];

  for (const toolCall of toolCalls) {
    const funcName = toolCall.function.name;
    const funcArgsStr = toolCall.function.arguments;

    let funcArgs;
    try {
      funcArgs = JSON.parse(funcArgsStr);
    } catch {
      funcArgs = {};
      logger.warn('processToolCalls: failed to parse arguments', { funcName });
    }

    const toolResult = executeTool(funcName, funcArgs);

    results.push({
      id: toolCall.id,
      function: funcName,
      arguments: funcArgs,
      result: toolResult,
    });
  }

  logger.info('processToolCalls: processed', { count: results.length });
  return results;
}

/**
 * Get list of registered tool names.
 * @returns {Array<string>} Tool names
 */
export function getAvailableTools() {
  return [...toolRegistry.keys()];
}

export default {
  WEATHER_TOOL,
  CALCULATOR_TOOL,
  DEFAULT_TOOLS,
  executeWeather,
  executeCalculate,
  registerTool,
  executeTool,
  processToolCalls,
  getAvailableTools,
};
